File: backend/app/schemas/tools.py
from typing import Any, Literal

from pydantic import BaseModel, Field


class ToolMetaResponse(BaseModel):
    tool_id: str
    name: str
    description: str
    input_types: list[str] = Field(default_factory=list)
    requires_file: bool = True
    supports_ai: bool = False
    actions: list[str] = Field(default_factory=list)
    enabled: bool = True
    disabled_title: str | None = None
    disabled_message: str | None = None


class ToolExecutionRequest(BaseModel):
    file_id: str | None = None
    params: dict[str, Any] = Field(default_factory=dict)


class ToolRunResponse(BaseModel):
    tool_id: str
    file_id: str | None = None
    result: dict[str, Any]


class ToolActionResponse(BaseModel):
    tool_id: str
    action: str
    file_id: str | None = None
    result: Any


class SqliteTableColumnResponse(BaseModel):
    name: str
    type: str
    not_null: bool
    default_value: str | None = None
    is_primary_key: bool = False


class SqliteTableInfoResponse(BaseModel):
    table_name: str
    row_count: int
    columns: list[SqliteTableColumnResponse] = Field(default_factory=list)
    schema_sql: str | None = None


class SqliteBrowserResponse(BaseModel):
    file_id: str
    database_name: str
    tables: list[SqliteTableInfoResponse] = Field(default_factory=list)


class SqlitePreviewFilterRequest(BaseModel):
    column: str = Field(min_length=1, max_length=255)
    operator: Literal["contains", "equals", "starts_with", "ends_with", "gt", "gte", "lt", "lte", "is_null", "not_null"] = "contains"
    value: str | None = None


class SqlitePreviewRequest(BaseModel):
    table_name: str = Field(min_length=1, max_length=255)
    selected_columns: list[str] = Field(default_factory=list)
    filters: list[SqlitePreviewFilterRequest] = Field(default_factory=list)
    limit: int = Field(default=50, ge=1, le=200)
    offset: int = Field(default=0, ge=0)


class SqlitePreviewResponse(BaseModel):
    table_name: str
    selected_columns: list[str] = Field(default_factory=list)
    available_columns: list[SqliteTableColumnResponse] = Field(default_factory=list)
    rows: list[dict[str, Any]] = Field(default_factory=list)
    total_rows: int
    returned_rows: int


class SqliteExportRequest(BaseModel):
    table_name: str = Field(min_length=1, max_length=255)
    selected_columns: list[str] = Field(default_factory=list)
    filters: list[SqlitePreviewFilterRequest] = Field(default_factory=list)
    include_header: bool = True
    delimiter: Literal[",", ";", "\\t", "|"] = ","


class SqliteExportResponse(BaseModel):
    table_name: str
    csv_name: str
    csv_url: str
    row_count: int
    columns: list[str] = Field(default_factory=list)


class AIStatusResponse(BaseModel):
    configured: bool
    chat_model: str | None = None
    reasoner_model: str | None = None


class ToolAIRequest(BaseModel):
    tool_id: str = Field(min_length=1, max_length=100)
    user_input: str = Field(min_length=1, max_length=8000)
    mode: Literal["chat", "reasoner"] = "chat"
    file_id: str | None = None
    context: dict[str, Any] = Field(default_factory=dict)


class ToolAIResponse(BaseModel):
    tool_id: str
    mode: Literal["chat", "reasoner"]
    model: str | None = None
    source: Literal["ai", "fallback"]
    reasoning: str = ""
    result: dict[str, Any]


class HashcatHashModeResponse(BaseModel):
    mode: int
    name: str
    category: str
    label: str


class HashcatTaskStatusResponse(BaseModel):
    enabled: bool = True
    disabled_title: str | None = None
    disabled_message: str | None = None
    configured: bool
    binary_path: str | None = None
    binary_source: str | None = None
    detected_platform: str
    bundle_dir: str | None = None
    bundled_binary_path: str | None = None
    wordlists_dir: str | None = None
    runtime_dir: str | None = None
    default_wordlist_path: str | None = None
    default_wordlist_name: str | None = None
    running: bool
    task_id: str | None = None
    pid: int | None = None
    command: list[str] = Field(default_factory=list)
    started_at: str | None = None
    finished_at: str | None = None
    exit_code: int | None = None
    hash_file: str | None = None
    result_file: str | None = None
    result_lines: list[str] = Field(default_factory=list)
    output_tail: list[str] = Field(default_factory=list)

File: frontend/src/types/tools.ts
export interface ToolMeta {
  tool_id: string
  name: string
  description: string
  input_types: string[]
  requires_file: boolean
  supports_ai: boolean
  actions: string[]
  enabled: boolean
  disabled_title: string | null
  disabled_message: string | null
}

export interface HashToolResult {
  file_name: string
  file_size: number
  algorithms: string[]
  available_algorithms?: string[]
  hashes: Record<string, string>
}

export interface SqliteCsvExportFile {
  table_name: string
  row_count: number
  columns: string[]
  csv_name: string
  csv_url: string
}

export interface SqliteCsvResult {
  database_name: string
  table_count: number
  tables: SqliteCsvExportFile[]
  zip_name: string
  zip_url: string
}

export interface SqliteColumnInfo {
  name: string
  type: string
  not_null: boolean
  default_value: string | null
  is_primary_key: boolean
}

export interface SqliteTableInfo {
  table_name: string
  row_count: number
  columns: SqliteColumnInfo[]
  schema_sql: string | null
}

export interface SqliteBrowserResult {
  file_id: string
  database_name: string
  tables: SqliteTableInfo[]
}

export type SqliteFilterOperator =
  | 'contains'
  | 'equals'
  | 'starts_with'
  | 'ends_with'
  | 'gt'
  | 'gte'
  | 'lt'
  | 'lte'
  | 'is_null'
  | 'not_null'

export interface SqlitePreviewFilter {
  column: string
  operator: SqliteFilterOperator
  value?: string | null
}

export interface SqlitePreviewResult {
  table_name: string
  selected_columns: string[]
  available_columns: SqliteColumnInfo[]
  rows: Record<string, unknown>[]
  total_rows: number
  returned_rows: number
}

export interface SqliteExportTableResult {
  table_name: string
  csv_name: string
  csv_url: string
  row_count: number
  columns: string[]
}

export interface TimestampParserParams {
  timestamp: string
  timestamp_type: string
  origin_timezone: string
  target_timezone: string
}

export interface TimestampParserResult {
  timestamp: string
  timestamp_type: string
  timestamp_type_label: string
  origin_timezone: string
  target_timezone: string
  converted_time: string
  supported_timezones: string[]
}

export interface TimestampAIAssistResult {
  timestamp: string
  timestamp_type: string
  origin_timezone: string
  target_timezone: string
  explanation: string
  confidence: string
  warnings: string[]
}

export interface ToolRunResponse {
  tool_id: string
  file_id: string | null
  result: Record<string, unknown>
}

export type HashcatAttackMode = 0 | 1 | 3 | 6 | 7

export interface HashcatHashMode {
  mode: number
  name: string
  category: string
  label: string
}

export interface HashcatTaskStatus {
  enabled: boolean
  disabled_title: string | null
  disabled_message: string | null
  configured: boolean
  binary_path: string | null
  binary_source: string | null
  detected_platform: string
  bundle_dir: string | null
  bundled_binary_path: string | null
  wordlists_dir: string | null
  runtime_dir: string | null
  default_wordlist_path: string | null
  default_wordlist_name: string | null
  running: boolean
  task_id: string | null
  pid: number | null
  command: string[]
  started_at: string | null
  finished_at: string | null
  exit_code: number | null
  hash_file: string | null
  result_file: string | null
  result_lines: string[]
  output_tail: string[]
}

export interface HashcatAIAssistResult {
  hash_mode: number
  attack_mode: HashcatAttackMode
  wordlist_path: string | null
  secondary_wordlist_path: string | null
  mask: string | null
  session_name: string | null
  extra_args: string[]
  explanation: string
  confidence: string
  warnings: string[]
}
